File: SRM 694/SRMDiv0Easy.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>
#include <queue>

using namespace std;
#define REP(i,n) for((i)=0;(i)<(int)(n);(i)++) 

typedef int F;
#define F_INF (1<<29)
#define MAXV 300
#define MAXE 10000

F cap[MAXE], flow[MAXE];
int to[MAXE], _prev[MAXE], last[MAXV], used[MAXV], level[MAXV];
struct MaxFlow{
    int V, E;
    MaxFlow(int n) {
        int i;
        V = n; E = 0;
        REP(i, V) last[i] = -1;
    }

    void add_edge(int x, int y, F f) {
        cap[E] = f; flow[E] = 0; to[E] = y; _prev[E] = last[x]; last[x] = E; E++;
        cap[E] = 0; flow[E] = 0; to[E] = x; _prev[E] = last[y]; last[y] = E; E++;
    }

    bool bfs(int s, int t) {
        int i;
        REP(i,V) level[i] = -1;
        queue <int> q;
        q.push(s); level[s] = 0;
        while (!q.empty()) {
            int x = q.front(); q.pop();
            for (i = last[x]; i >= 0; i = _prev[i]) if(level[to[i]] == -1 && cap[i] > flow[i]) {
                q.push(to[i]); level[to[i]] = level[x] + 1;
            }
        }
        return (level[t] != -1);
    }
    F dfs(int v, int t, F f) {
        int i;
        if (v == t) return f;
        for (i = used[v]; i >= 0; used[v]=i=_prev[i]) if(level[to[i]] > level[v] && cap[i] >flow[i]) {
            F tmp = dfs(to[i], t, min(f, cap[i] - flow[i]));
            if (tmp > 0) {flow[i] += tmp; flow[i^1] -= tmp; return tmp;}
        }
        return 0;
    }
    F maxflow(int s, int t) {
        int i;
        while(bfs(s,t)) {
            REP(i,V) used[i] = last[i];
            while (dfs(s, t, F_INF) != 0);
        }
        F ans = 0;
        for (i=last[s]; i >= 0; i = _prev[i]) ans += flow[i];
        return ans;
    }
};
int delta[210];
class SRMDiv0Easy {
    public:
    int get(int N, vector<int> L, vector<int> R, vector<int> X, vector<int> Y) {
        int i;
        MaxFlow mf(N+3);
        mf.add_edge(N+1, 0, 0);
        mf.add_edge(N, N+2, 0);
        REP(i,L.size()) {
            mf.add_edge(L[i], R[i]+1, Y[i]-X[i]);
            delta[L[i]] += X[i];
            delta[R[i] + 1] -= X[i];
        }
        int S = 0;
        REP(i,N+1) {
            if (delta[i] > 0) {
                mf.add_edge(i, N+2, delta[i]);
                S += delta[i];
            } else if (delta[i] < 0) {
                mf.add_edge(N+1, i, -delta[i]);
            }
        }
        int ans = -1;
        REP(i,200010){
            if(mf.maxflow(N+1,N+2) == S + i) ans = i;
            cap[0]++;
            cap[2]++;
        }
        return ans;
    }
};

// CUT begin
ifstream data("SRMDiv0Easy.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(int N, vector<int> L, vector<int> R, vector<int> X, vector<int> Y, int __expected) {
    time_t startClock = clock();
    SRMDiv0Easy *instance = new SRMDiv0Easy();
    int __result = instance->get(N, L, R, X, Y);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        int N;
        from_stream(N);
        vector<int> L;
        from_stream(L);
        vector<int> R;
        from_stream(R);
        vector<int> X;
        from_stream(X);
        vector<int> Y;
        from_stream(Y);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(N, L, R, X, Y, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1468737157;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 900 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "SRMDiv0Easy (900 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end

File: SRM 694/SRMDiv0Easy.java
import java.io.*;
import java.util.*;

public class SRMDiv0Easy {
	private static int curGen = 0;
	private static final int INF = (int) 1e9;

	SRMDiv0Easy.Edge addEdge(SRMDiv0Easy.Vertex a, SRMDiv0Easy.Vertex b, int cap) {
		SRMDiv0Easy.Edge ab = new SRMDiv0Easy.Edge();
		SRMDiv0Easy.Edge ba = new SRMDiv0Easy.Edge();
		ab.dest = b; ab.cap = cap; ab.flow = 0; ab.rev = ba;
		ba.dest = a; ba.cap = 0; ba.flow = 0; ba.rev = ab;
		a.outgo.add(ab);
		a.outgo.add(ba);
		return ab;
	}

	public int get(int N, int[] L, int[] R, int[] X, int[] Y) {
		SRMDiv0Easy.Vertex[] vs = new SRMDiv0Easy.Vertex[N + 1];
		for (int i = 0; i <= N; ++i) {
			vs[i] = new SRMDiv0Easy.Vertex();
		}
		int[] delta = new int[N + 1];
		SRMDiv0Easy.Edge[] direct = new SRMDiv0Easy.Edge[L.length];
		for (int i = 0; i < L.length; ++i) {
			direct[i] = addEdge(vs[L[i]], vs[R[i] + 1], Y[i] - X[i]);
			delta[L[i]] -= X[i];
			delta[R[i] + 1] += X[i];
		}
		SRMDiv0Easy.Edge loop = addEdge(vs[N], vs[0], INF);
		SRMDiv0Easy.Vertex s = new SRMDiv0Easy.Vertex();
		SRMDiv0Easy.Vertex t = new SRMDiv0Easy.Vertex();
		int need = 0;
		SRMDiv0Easy.Edge[] aux = new SRMDiv0Easy.Edge[N + 1];
		for (int i = 0; i <= N; ++i) {
			if (delta[i] < 0) {
				aux[i] = addEdge(vs[i], t, -delta[i]);
			} else {
				aux[i] = addEdge(s, vs[i], delta[i]);
			}
			need += Math.abs(delta[i]);
		}
		if (maxFlow(s, t) != need / 2) {
			return -1;
		}
		int res = loop.flow;
		loop.cap = loop.rev.cap = 0;
		loop.flow = loop.rev.flow = 0;
		for (SRMDiv0Easy.Edge e : aux) {
			e.cap = e.rev.cap = 0;
			e.flow = e.rev.flow = 0;
		}
		for (int i = 0; i < L.length; ++i) {
			direct[i].cap += X[i];
			direct[i].flow += X[i];
			direct[i].rev.cap -= X[i];
			direct[i].rev.cap -= X[i];
		}
		res += maxFlow(vs[0], vs[N]);
		return res;
	}

	private int maxFlow(SRMDiv0Easy.Vertex s, SRMDiv0Easy.Vertex t) {
		int res = 0;
		while(true) {
			++curGen;
			int by = s.dfs(t, INF);
			if (by == 0) break;
			res += by;
		}
		return res;
	}

	static class Edge {
		SRMDiv0Easy.Vertex dest;
		int cap;
		int flow;
		SRMDiv0Easy.Edge rev;
	}

	static class Vertex {
		int gen = 0;
		List<SRMDiv0Easy.Edge> outgo = new ArrayList<Edge>(1);
		int ptr = 0;
		public int dfs(SRMDiv0Easy.Vertex t, int max) {
			if (this == t) return max;
			if (gen == curGen) return 0;
			gen = curGen;
			for (int i = 0; i < outgo.size(); ++i) {
				SRMDiv0Easy.Edge e = outgo.get(ptr);
				if (e.flow < e.cap) {
					int nby = e.dest.dfs(t, Math.min(max, e.cap - e.flow));
					if (nby != 0) {
						e.flow += nby;
						e.rev.flow -= nby;
						return nby;
					}
				}
				++ptr;
				if (ptr >= outgo.size()) ptr = 0;
			}
			return 0;
		}
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("SRMDiv0Easy (900 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int N = Integer.parseInt(Reader.nextLine());
            int[] L = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < L.length; ++i)
                L[i] = Integer.parseInt(Reader.nextLine());
            int[] R = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < R.length; ++i)
                R[i] = Integer.parseInt(Reader.nextLine());
            int[] X = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < X.length; ++i)
                X[i] = Integer.parseInt(Reader.nextLine());
            int[] Y = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < Y.length; ++i)
                Y[i] = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(N, L, R, X, Y, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1468768717;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 900 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int N, int[] L, int[] R, int[] X, int[] Y, int __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		SRMDiv0Easy instance = new SRMDiv0Easy();
		int __result = 0;
		try {
			__result = instance.get(N, L, R, X, Y);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "/Users/resry/topcoder/SRM 694/SRMDiv0Easy.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}
